File: src/main/java/retouch/project/careNdShare/controller/DonateController.java
package retouch.project.careNdShare.controller;

import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import retouch.project.careNdShare.dto.DonateItemDTO;
import retouch.project.careNdShare.dto.DonateRequestDTO;
import retouch.project.careNdShare.service.DonateItemService;
import retouch.project.careNdShare.service.DonateRequestService;

@RestController
@RequestMapping("/api/donate")
@CrossOrigin(origins = "*")
public class DonateController {

    @Autowired
    private DonateItemService donateItemService;

    @Autowired
    private DonateRequestService donateRequestService;

    @PostMapping(value = "/add", consumes = "multipart/form-data")
    public ResponseEntity<?> addDonation(@Valid @ModelAttribute DonateItemDTO dto) {
        try {
            return ResponseEntity.ok(donateItemService.addDonation(dto));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/available")
    public ResponseEntity<?> getAvailableDonations(@RequestParam(required = false) String type) {
        try {
            return ResponseEntity.ok(donateItemService.getAvailableDonations(type));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/my-donations")
    public ResponseEntity<?> getMyDonations() {
        try {
            return ResponseEntity.ok(donateItemService.getMyDonations());
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getDonationById(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(donateItemService.getDonationById(id));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    // ✅ ADD THIS - for View Details feature
    @GetMapping("/details/{id}")
    public ResponseEntity<?> getDonationDetails(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(donateItemService.getDonationById(id));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteDonation(@PathVariable Long id) {
        try {
            donateItemService.deleteDonation(id);
            return ResponseEntity.ok("Donation deleted successfully");
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @PostMapping("/request")
    public ResponseEntity<?> createRequest(@Valid @RequestBody DonateRequestDTO dto) {
        try {
            return ResponseEntity.ok(donateRequestService.createRequest(dto));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/my-requests")
    public ResponseEntity<?> getMyRequests() {
        try {
            return ResponseEntity.ok(donateRequestService.getMyRequests());
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }
}

File: controllers/donate.controller.js
const donateItemService = require('../services/donateItem.service');
const donateRequestService = require('../services/donateRequest.service');

class DonateController {
  // Add donation
  async addDonation(req, res) {
    try {
      const donationData = {
        itemType: req.body.itemType,
        itemName: req.body.itemName,
        quantity: parseInt(req.body.quantity),
        itemCondition: req.body.itemCondition,
        pickupAddress: req.body.pickupAddress
      };

      const files = req.files;

      const donation = await donateItemService.addDonation(donationData, files, req.user.id);

      return res.json(donation);
    } catch (error) {
      console.error('Add donation error:', error);
      return res.status(400).json({ message: error.message });
    }
  }

  // Get available donations
  async getAvailableDonations(req, res) {
    try {
      const { type } = req.query;
      const donations = await donateItemService.getAvailableDonations(type);
      return res.json(donations);
    } catch (error) {
      console.error('Get available donations error:', error);
      return res.status(400).json({ message: error.message });
    }
  }

  // Get my donations
  async getMyDonations(req, res) {
    try {
      const donations = await donateItemService.getMyDonations(req.user.id);
      return res.json(donations);
    } catch (error) {
      console.error('Get my donations error:', error);
      return res.status(400).json({ message: error.message });
    }
  }

  // Get donation by ID
  async getDonationById(req, res) {
    try {
      const { id } = req.params;
      const donation = await donateItemService.getDonationById(id);

      if (!donation) {
        return res.status(404).json({ message: 'Donation not found' });
      }

      return res.json(donation);
    } catch (error) {
      console.error('Get donation error:', error);
      return res.status(400).json({ message: error.message });
    }
  }

  // Delete donation
  async deleteDonation(req, res) {
    try {
      const { id } = req.params;
      await donateItemService.deleteDonation(id, req.user.id);
      return res.json({ message: 'Donation deleted successfully' });
    } catch (error) {
      console.error('Delete donation error:', error);
      return res.status(400).json({ message: error.message });
    }
  }

  // Create donation request
  async createRequest(req, res) {
    try {
      const requestData = {
        donationId: req.body.donationId,
        description: req.body.description
      };

      const request = await donateRequestService.createRequest(requestData, req.user.id);
      return res.json(request);
    } catch (error) {
      console.error('Create request error:', error);
      return res.status(400).json({ message: error.message });
    }
  }

  // Get my donation requests
  async getMyRequests(req, res) {
    try {
      const requests = await donateRequestService.getMyRequests(req.user.id);
      return res.json(requests);
    } catch (error) {
      console.error('Get my requests error:', error);
      return res.status(400).json({ message: error.message });
    }
  }
}

module.exports = new DonateController();
